src/Internal/BalanceTx/Types.purs

```diff
@@ -1,22 +1,78 @@
 module Ctl.Internal.BalanceTx.Types
   ( BalanceTxM
+  , BalanceTxMContext
   , FinalizedTransaction(FinalizedTransaction)
   , PrebalancedTransaction(PrebalancedTransaction)
+  , askCoinsPerUtxoUnit
+  , askConstraints
+  , liftEitherQueryM
+  , liftQueryM
+  , withBalanceTxConstraints
   ) where
 
 import Prelude
 
-import Control.Monad.Except.Trans (ExceptT)
+import Control.Monad.Except.Trans (ExceptT(ExceptT))
+import Control.Monad.Reader.Trans (asks, withReaderT)
+import Control.Monad.Trans.Class (lift)
+import Ctl.Internal.BalanceTx.Constraints
+  ( BalanceTxConstraints
+  , BalanceTxConstraintsBuilder
+  )
+import Ctl.Internal.BalanceTx.Constraints
+  ( buildBalanceTxConstraints
+  ) as Constraints
 import Ctl.Internal.BalanceTx.Error (BalanceTxError)
 import Ctl.Internal.Cardano.Types.Transaction (Transaction)
-import Ctl.Internal.QueryM (QueryMExtended)
+import Ctl.Internal.QueryM
+  ( DefaultQueryEnv
+  , QueryEnv
+  , QueryM
+  , QueryMExtended(QueryMExtended)
+  )
+import Ctl.Internal.QueryM (liftQueryM) as QueryM
+import Ctl.Internal.QueryM.Ogmios (CoinsPerUtxoUnit)
 import Ctl.Internal.Types.ScriptLookups (UnattachedUnbalancedTx)
+import Data.Either (Either)
 import Data.Generic.Rep (class Generic)
-import Data.Newtype (class Newtype)
+import Data.Newtype (class Newtype, over, unwrap)
 import Data.Show.Generic (genericShow)
 import Effect.Aff (Aff)
 
-type BalanceTxM (a :: Type) = ExceptT BalanceTxError (QueryMExtended () Aff) a
+type BalanceTxMContext = (constraints :: BalanceTxConstraints)
+
+type BalanceTxM (a :: Type) =
+  ExceptT BalanceTxError (QueryMExtended BalanceTxMContext Aff) a
+
+askCoinsPerUtxoUnit :: BalanceTxM CoinsPerUtxoUnit
+askCoinsPerUtxoUnit =
+  asks (_.coinsPerUtxoUnit <<< unwrap <<< _.pparams <<< _.runtime)
+
+askConstraints :: BalanceTxM BalanceTxConstraints
+askConstraints = asks (_.constraints <<< _.extraConfig)
+
+liftQueryM :: forall (a :: Type). QueryM a -> BalanceTxM a
+liftQueryM = lift <<< QueryM.liftQueryM
+
+liftEitherQueryM
+  :: forall (a :: Type). QueryM (Either BalanceTxError a) -> BalanceTxM a
+liftEitherQueryM = ExceptT <<< QueryM.liftQueryM
+
+withBalanceTxConstraints
+  :: forall (a :: Type)
+   . BalanceTxConstraintsBuilder
+  -> QueryMExtended BalanceTxMContext Aff a
+  -> QueryM a
+withBalanceTxConstraints constraintsBuilder =
+  over QueryMExtended (withReaderT setQueryEnv)
+  where
+  setQueryEnv :: DefaultQueryEnv -> QueryEnv BalanceTxMContext
+  setQueryEnv defaultEnv =
+    let
+      constraints :: BalanceTxConstraints
+      constraints = Constraints.buildBalanceTxConstraints constraintsBuilder
+    in
+      defaultEnv { extraConfig = { constraints } }
```

Comment:
This is perhaps the only part of the implementation I might request changing. We're going to remove `extraConfig` in this upcoming version, so it would probably be a good idea to future-proof this a bit and choose another way of passing the constraints context (just another `ReaderT`?)

Reply:
This is a good point, I updated `BalanceTxM` to use a separate `ReaderT` for the constraints context